ape_safe/factory.py

```diff
@@ -0,0 +1,115 @@
+import secrets
+from functools import cache
+from typing import TYPE_CHECKING, Iterable, Optional, Union
+
+from ape.types import AddressType
+from ape.utils import ZERO_ADDRESS, ManagerAccessMixin
+from packaging.version import Version
+
+from ape_safe.packages import SAFE_PACKAGE_BY_VERSION, PackageType, get_factory, get_singleton
+
+if TYPE_CHECKING:
+    from ape.api import AccountAPI
+    from ape.contracts import ContractInstance
+
+
+class Factory(ManagerAccessMixin):
+    _singleton: dict[Version, "ContractInstance"] = {}
+    _factory: dict[Version, "ContractInstance"] = {}
+
+    @classmethod
+    def inject(cls, version: Version, deployer: "AccountAPI"):
+        cls._singleton[version] = deployer.deploy(PackageType.SINGLETON(version))
+        cls._factory[version] = deployer.deploy(PackageType.PROXY_FACTORY(version))
+
+    @cache
+    def get_factory(self, version: Version) -> "ContractInstance":
+        if injected_factory := self._factory.get(version):
+            return injected_factory
+
+        return get_factory(self.chain_manager.chain_id, version)
+
+    @property
+    def contract(self) -> "ContractInstance":
+        return self.get_factory(max(SAFE_PACKAGE_BY_VERSION))
+
+    @cache
+    def get_singleton(self, version: Version) -> "ContractInstance":
+        if injected_singleton := self._singleton.get(version):
+            return injected_singleton
+
+        return get_singleton(self.chain_manager.chain_id, version)
+
+    def create(
+        self,
+        owners: Iterable[Union["AddressType", str]],
+        threshold: int,
+        callback_address: Union["AddressType", str] = ZERO_ADDRESS,
+        callback_calldata: Optional[bytes] = None,
+        fallback_handler: Union["AddressType", str] = ZERO_ADDRESS,
+        payment_token: Union["AddressType", str] = ZERO_ADDRESS,
+        payment_amount: Union[str, int] = 0,
+        payment_receiver: Union["AddressType", str] = ZERO_ADDRESS,
+        salt: Optional[int] = None,
+        version: Union[Version, str, None] = None,
+        **txn_kwargs,
+    ) -> "ContractInstance":
+        if not (owners := [self.conversion_manager.convert(a, AddressType) for a in owners]):
+            raise ValueError("Cannot make a Safe with 0 owners.")
+
+        elif not (1 <= threshold <= len(owners)):
+            raise ValueError(f"Threshold must be between '1' and '{len(owners)}'")
+
+        if callback_address != ZERO_ADDRESS:
+            callback_address = self.conversion_manager.convert(callback_address, AddressType)
+
+        if fallback_handler != ZERO_ADDRESS:
+            fallback_handler = self.conversion_manager.convert(fallback_handler, AddressType)
+
+        if (payment_amount := self.conversion_manager.convert(payment_amount, int)) > 0 and (
+            payment_token == ZERO_ADDRESS or payment_receiver == ZERO_ADDRESS
+        ):
+            raise ValueError
```

Comment:
You can't raise the class, it has to be an instance.

```suggestion
            raise ValueError()
```

Reply:
156a6d3